### PR TITLE
fix(array): validate run-end encoded invariants

### PR DESCRIPTION
**Pull request by fallintoplace**
## Summary
- add RunEndEncoded Validate and ValidateFull implementations
- validate parent and child invariants, logical coverage, and strict run-end monotonicity
- add regression tests for short coverage and duplicate run ends

## Why
Run-end encoded helpers already assume that run ends are strictly increasing and cover the logical range. Today malformed arrays can pass validation even when those invariants do not hold.

## Validation
- go test ./arrow/array ./arrow/encoded

### CODE REVIEW

File: arrow/array/encoded.go [hunk 1]
```diff
@@ -240,6 +249,102 @@ func (r *RunEndEncoded) String() string {
...
+
+	runEnds := getRunEnds64(runEndsData)
+	lastRunEnd := runEnds[len(runEnds)-1]
+	if lastRunEnd < int64(r.data.offset+r.data.length) {
+		return fmt.Errorf("arrow/array: last run end is %d but it should cover %d", lastRunEnd, r.data.offset+r.data.length)
+	}
+
+	if !full {
+		return nil
+	}
```
Comment (zeroshade) [inline arrow/array/encoded.go]: In the `!full` path, we're copying the entirety of the run ends only to read the single `runEnds[len-1]`. Instead for the `!full` path we should *only* check the last run end instead of doing the full copy loop

File: arrow/array/encoded.go [hunk 2]
```diff
@@ -240,6 +249,102 @@ func (r *RunEndEncoded) String() string {
...
+		runEnds = runEnds[data.Offset() : data.Offset()+data.Len()]
+		out := make([]int64, len(runEnds))
+		for i, v := range runEnds {
+			out[i] = int64(v)
+		}
+		return out
+	default:
+		runEnds := arrow.Int64Traits.CastFromBytes(data.Buffers()[1].Bytes())
+		return runEnds[data.Offset() : data.Offset()+data.Len()]
+	}
```
Comment (zeroshade) [inline arrow/array/encoded.go]: at this point we haven't validated the length/buffers yet, so we need to check `len(runEnds)` before you start slicing using `data.Offset()+data.Len()` otherwise we could panic on a corrupt buffer.

File: arrow/array/encoded.go [hunk 3]
```diff
@@ -240,6 +249,102 @@ func (r *RunEndEncoded) String() string {
...
+	if len(data.buffers) > 0 && data.buffers[0] != nil {
+		return data.length - bitutil.CountSetBits(data.buffers[0].Bytes(), data.offset, data.length)
+	}
+	return 0
+}
+
+func runEndTypeLimit(id arrow.Type) int64 {
+	switch id {
+	case arrow.INT16:
+		return 1<<15 - 1
```
Comment (zeroshade) [inline arrow/array/encoded.go]: `math.MaxInt16` ?

File: arrow/array/encoded.go [hunk 4]
```diff
@@ -240,6 +249,102 @@ func (r *RunEndEncoded) String() string {
...
+	case arrow.INT32:
+		return math.MaxInt32
+	default:
+		return math.MaxInt64
+	}
+}
+
+func getRunEnds64(data arrow.ArrayData) []int64 {
```
Comment (zeroshade) [inline arrow/array/encoded.go]: there's already a `encoded.getRunEnds` function at `ree_utils.go:109` which returns a non-allocating closure, could we use export/reuse that pattern instead of allocating a new []int64?

**Review by zeroshade — CHANGES_REQUESTED** [review]
Can we add a test that goes through the `array.Validate`/`array.ValidateFull` functions also?

**Review by zeroshade — APPROVED** [review]
